### PR TITLE
Build transaction with token burning

### PR DESCRIPTION
**Pull request by aslesarenko**
Required for https://github.com/ergoplatform/ergo-appkit/issues/51

### CODE REVIEW

File: ergo-wallet/src/main/scala/org/ergoplatform/wallet/AssetUtils.scala [hunk 1]
```diff
@@ -7,22 +7,41 @@ import scala.collection.mutable
...
+      }
+    }
+  }
+
+  @inline
+  def subtractAssets(from: TokensMap, subtractor: TokensMap*): TokensMap = {
+    subtractor.foldLeft(from){ case (from, substractor) =>
+      substractor.foldLeft(from) { case (acc, (id, amount)) =>
+        val currAmt = acc.getOrElse(id,
+          throw new IllegalArgumentException(s"Cannot subtract $amount of token $id: token not found in $acc"))
```
Comment (kushti) [inline ergo-wallet/src/main/scala/org/ergoplatform/wallet/AssetUtils.scala:31]: Why checks are different from "subtractAssetsMut" method? Please unify.

File: ergo-wallet/src/main/scala/org/ergoplatform/wallet/transactions/TransactionBuilder.scala [hunk 2]
```diff
@@ -92,10 +93,15 @@ object TransactionBuilder {
     val tokensOutNoMinted = tokensOut.filterKeys(_ != firstInputBoxId)
     val mintedTokensNum = tokensOut.size - tokensOutNoMinted.size
     require(mintedTokensNum <= 1, s"Only one token can be minted, but found $mintedTokensNum")
+    require(burnTokens.values.forall(_ > 0), s"Incorrect burnTokens specification: $burnTokens")
```
Comment (kushti) [inline ergo-wallet/src/main/scala/org/ergoplatform/wallet/transactions/TransactionBuilder.scala:96]: While other checks are protocol specific, this particular is not, Please make a comment about that.

File: ergo-wallet/src/test/scala/org/ergoplatform/wallet/utils/WalletTestHelpers.scala [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+          fail(s"exception check failed on $e (root cause: ${rootCause(e)}) \n trace:\n${e.getStackTrace.mkString("\n")}}")
+    }
+  }
+
+  @tailrec
+  final def rootCause(t: Throwable): Throwable =
+    if (t.getCause == null) t
+    else rootCause(t.getCause)
+
+  def tid(s: String): ModifierId = {
```
[the remark this answers is not in the document]
Comment (aslesarenko) [reply ergo-wallet/src/test/scala/org/ergoplatform/wallet/utils/WalletTestHelpers.scala:28]: done.

File: ergo-wallet/src/main/scala/org/ergoplatform/wallet/AssetUtils.scala [hunk 4]
```diff
@@ -7,22 +7,41 @@ import scala.collection.mutable
...
+    from.foldLeft(to){ case (to, from) =>
+      from.foldLeft(to) { case (acc, (id, amount)) =>
+        acc.updated(id, Math.addExact(acc.getOrElse(id, 0L), amount))
+      }
+    }
+  }
+
+  @inline
+  def subtractAssets(from: TokensMap, subtractor: TokensMap*): TokensMap = {
+    subtractor.foldLeft(from){ case (from, substractor) =>
```
[the remark this answers is not in the document]
Comment (aslesarenko) [reply ergo-wallet/src/main/scala/org/ergoplatform/wallet/AssetUtils.scala:28]: Done

File: ergo-wallet/src/main/scala/org/ergoplatform/wallet/AssetUtils.scala [hunk 5]
```diff
@@ -7,22 +7,41 @@ import scala.collection.mutable
...
+        into.put(id, Math.addExact(into.getOrElse(id, 0L), amount))
     })
   }
 
   @inline
-  def mergeAssets(from: Map[ModifierId, Long], to: Map[ModifierId, Long] = Map.empty): Map[ModifierId, Long] = {
-    from.foldLeft(to) { case (acc, (id, amount)) =>
-      acc.updated(id, acc.getOrElse(id, 0L) + amount)
+  def mergeAssets(to: TokensMap, from: TokensMap*): TokensMap = {
+    from.foldLeft(to){ case (to, from) =>
```
[the remark this answers is not in the document]
Comment (aslesarenko) [reply ergo-wallet/src/main/scala/org/ergoplatform/wallet/AssetUtils.scala:19]: Done

File: ergo-wallet/src/main/scala/org/ergoplatform/wallet/AssetUtils.scala (hunk 1 above)
Comment (aslesarenko) [reply ergo-wallet/src/main/scala/org/ergoplatform/wallet/AssetUtils.scala:31]: @kushti subtractAssetsMut is used in DefaultBoxSelector.formChangeBoxes, which in turn is used in applications. Adding this checks will make the method more strict. 
Can do it, as long as it is not a big deal.

File: ergo-wallet/src/main/scala/org/ergoplatform/wallet/transactions/TransactionBuilder.scala (hunk 2 above)
Comment (aslesarenko) [reply ergo-wallet/src/main/scala/org/ergoplatform/wallet/transactions/TransactionBuilder.scala:96]: done

**Pull request merged**